storybookjs/storybook
# WIP Add support for idiomatic svelte3 stories

Issue: https://github.com/storybookjs/storybook/issues/6653

## What I did

- added a new  package in [app/svelte3](https://github.com/rixo/storybook/tree/svelte3/app/svelte3)

- copied existing svelte example to [example/svelte3-kitchen-sink](https://github.com/rixo/storybook/tree/svelte3/examples/svelte3-kitchen-sink) and adapted them to the new syntax

With the new package, stories are written directly as svelte components in `.svelte` files, as described in the linked issue. It supports both a [shortcut syntax](https://github.com/rixo/storybook/blob/svelte3/examples/svelte3-kitchen-sink/src/stories/addon-actions.stories.svelte) for files with a single kind of stories, and a [complete syntax](https://github.com/rixo/storybook/blob/svelte3/examples/svelte3-kitchen-sink/src/stories/syntax-context-storiesof.stories.svelte#L43), with a `StoriesOf` component, to support edge cases and more closely stick to usual storybook syntax.

I don't like having a new package for svelte but the codebases of the existing svelte app and my new svelte3 are much different and probably irreconcilable. Furthermore, existing package supports svelte v2, which my version won't be able to do. So I have no idea what should be the proper course of action here, so I preferred not to touch the existing package until I get your input.

## How to test

- Is this testable with Jest or Chromatic screenshots? **TODO**
- Does this need a new example in the kitchen sink apps? **DONE**
- Does this need an update to the documentation? **TODO**

See the `svelte3-kitchen-sink` example.

## Addons support

Examples for addons `actions`, `backgrounds`, `knobs`, `links`, and `notes` are included and functional.

An implementation of the `centered` addon is included in the app package because I didn't manage to find how to integrate with the actual `centered` package yet.

The `storysource` addon is not supported currently.

Other addons have not been tested.

## TODO

Tests, docs, some addons support...

Also, I think current compilation / webpack options are quite fragile. In particular, I had some interop issues with svelte compiler and the existing babel compilation. As a result, the `app/svelte3/dist` contains raw `.svelte` files that are only built in the (user's) storybook app; and svelte compiler does not transpile to ES5, so I'm pretty sure it won't work in IE11 and the like. I'll probably need assistance to fix that. I have had a pretty hard time with handling the build / repo, I must confess.
